### PR TITLE
Post/Put a post

### LINKED ISSUE #205: PostView: PUT

### PR DESCRIPTION
**Pull request by tylerelston**
Implemented 
- `PUT author/{AUTHOR_ID}/posts/{POST_ID}` (create post with post_id as id)
- `POST author/{AUTHOR_ID}/posts/` (create post)

Implemented as creation of LocalPosts
Resolves #205 #206

### CODE REVIEW

File: code/api/utility.py [hunk 1]
```diff
@@ -12,11 +13,55 @@ def getPaginated(data, page, size):
...
+            unlisted=data["unlisted"],
+    )
+
+    received_post.save()
+    # set post origin and source to itself for a new post
+    received_post.origin = received_post.get_id()
+    received_post.source = received_post.get_id()
+    if post_id:
+        received_post.id = post_id
+    received_post.save()
```
Comment (JawadHossain) [inline code/api/utility.py:37]: .save() called twice. lines 31 and 37

Comment (tylerelston) [reply code/api/utility.py:36]: Thanks, also removed a duplicate save in socialDistribution/views.py/posts

File: code/api/utility.py [hunk 2]
```diff
@@ -12,11 +13,55 @@ def getPaginated(data, page, size):
...
+            content_type=data["contentType"],
+            content=data["content"].encode('utf-8'),
+            visibility=LocalPost.Visibility.get_visibility_choice(data["visibility"]),
+            unlisted=data["unlisted"],
+    )
+
+    received_post.save()
+    # set post origin and source to itself for a new post
+    received_post.origin = received_post.get_id()
+    received_post.source = received_post.get_id()
```
Comment (mxiuwang) [inline code/api/utility.py:33]: nbd but does this need to be 2 lines

File: code/api/utility.py [hunk 3]
```diff
@@ -12,11 +13,55 @@ def getPaginated(data, page, size):
...
+        if category != "":
+            category_obj, cat_created = Category.objects.get_or_create(
+                category__iexact=category,
+                defaults={"category": category}
+            )
+            received_post.categories.add(category_obj)
+
+            # loop condition is always false if post was created
+            # because categories_to_remove is an empty list then
+            while category_obj.category in categories_to_remove:
```
Comment (mxiuwang) [inline code/api/utility.py:50]: is this nested for/while loops increase complexity? Can you just use if statement instead

Comment (tylerelston) [reply code/api/utility.py:50]: not sure. the category logic was written by someone in a different pr, we could refactor it in another pr if we want maybe?

Comment (mxiuwang) [reply code/api/utility.py:49]: sure. if it works it's nbd but just a thought

**Review by mxiuwang — APPROVED** [review]

**Comment by JawadHossain** [comment]
I cannot add images to posts. Can you pls check if it is the same for you?

**Comment by tylerelston** [comment]
> 

Hmm yeah it isn't working for me either, haven't been able to figure out why. I'll take another look tomorrow

**Comment by tylerelston** [comment]
> I cannot add images to posts. Can you pls check if it is the same for you?

Fixed it by setting the post's content type to our actual ContentType defined in the model, rather than just using the string from the request

File: code/socialDistribution/views.py [hunk 4]
```diff
@@ -437,6 +437,7 @@ def posts(request, author_id):
         form = PostForm(request.POST, request.FILES, user_id=user_id)
         if form.is_valid():
             content, content_type = form.get_content_and_type()
+            print(content)
```
Comment (JawadHossain) [inline code/socialDistribution/views.py:440]: print!

**Review by JawadHossain — APPROVED** [review]
Thanks for the fix!

**Pull request merged**